### ISSUE TITLE
Semantic Snippets

### ISSUE THREAD

**Issue by jmarolf** [dotnet/roslyn#56541]
Snippets as they exist today for .NET developers only do text expansion (cw -> Console.WriteLine) and have lacked any significant improvements in the last 10+ years. We would like to iterate on this experience for .NET developers using Visual Studio. Snippets based on the rich analysis capabilities of the roslyn api that can change based on context.

## Proposal

1. Do not show snippets in the completion list for C#
1. For the set of some 40 snippets we ship for the language replace them with "snippets" written using the completion provider API
1. Work with same teams as we do no roslyn-analyzer to write api-specific snippets and ship them in the SDK
1. Still ship snippets so muscle memory like `cw` -> `Console.WriteLine();` will work

### Advantages

![image](https://user-images.githubusercontent.com/9797472/133976691-ab13407e-25c0-427d-b443-af155af36ffc.png)

 The [CompletionProvider](https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/Completion/CompletionProvider.cs) api has lots of advanced capabilities that allow us to be semantically aware of what is around the snippet. I've been [playing around](https://github.com/jmarolf/smart-snippets) with what the API is capable of and there are a lot of possibilities.
- Only show completion for snippets when it is syntactically correct to do so
- Advanced filtering and sorting capabilities available to us via the completion api
- Populate values from members that are in scope (automatically include cancellation tokens etc.)
- Have different behavior depending on where the cursor is (surround with vs. expansion can just be where on the line you are typing)
- completion providers can come from nuget packages so we can have other teams such as the BCL 

### Open Implementation Questions

- In our own implementations we would use the Speculative Semantic Model and other tricks to ensure good performance, some of these helpers may need to get ported to like what was done for roslyn-analyzers if we want other teams to contribute
- We currently can create tab stop behavior by synthesizing an xml snippet on demand. Its unclear if this approach is something we want to continue doing.

### Open Design Questions

1. What should the pre-selection and filtering behavior be? If the snippets are targeted enough, it may be sufficient for their behavior to be simple.
1. If the user has third-party snippets installed, we should still show those but the ~40 we own should not appear
1. Potentially add a Tools->Option toggle to **show** snippets again and/or disable this feature. Its unclear how people will react to the old snippets not being in the completion list.

### C# Snippet Behavior in Visual Studio

<details><summary>List of all snippets C# included in Visual Studio</summary>
Here is a list of all 39 "Expansion" snippets we ship for C# today. These appear in the completion list as their snippet text entry.

![image](https://user-images.githubusercontent.com/9797472/133972704-ee49f95b-91f0-43dd-806c-f853511d541a.png)

| Snippet Text  |  Description |
|---------------|---------------|
|`#if` |  if region directing
|`#region` |  region directive
|`attribute` |  define new attribute class (25 lines)
|`checked` |  checked statement
|`class` |  define new class
|`ctor` |  define new parameter-less constructor
|`cw` |  Console.WriteLine
|`do` |  do…while loop
|`else` |  else statement
|`enum` |  define a new enum
|`equals` |  override `Equals(object obj)` and `GetHashCode() methods (27 lines)
|`exception`|  define a new exception type (10 lines)
|`for`|  for loop
|`foreach`|  foreach block
|`forr`|  Reverse for loop
|`if`|  if statement
|`indexer`|  indexer property
|`interface`|  define a new interface
|`iterindex`|  define an interator class with and index property (43 lines)
|`iterator` |  define `IEnumerator<ElementType> GetEnumerator()`method
|`lock`|  lock statement
|`mbox`|  show message box (`System.Windows.Forms.MessageBox.Show("Test");`)
|`namespace`| define a new namespace |
|`prop`| `public int MyProperty { get; set; }` |
|`propfull`| property with backing field (8 lines)
|`propg`|  Property with a get accessor and private set accessor
|`sim`|  int Main method
|`struct`| define a new struct
|`svm`|  void Main method
|`switch`|  switch statement
|`testc`|  define new MSTest Test Class
|`testm`|  define new MSTest Test Method
|`try`|  try catch
|`tryf`|  try finally
|`unchecked`|  unchecked block
|`unsafe`|  unsafe block
|`using`| using statement
|`while`|  while loop
|`~` |  deconstructor

Here is a list of all 21 "Surrounds With" snippets we ship for C# today. These require the user to select text and press te "Surround With" cord (Ctrl+K+S) and are never shown in the completion list.

![image](https://user-images.githubusercontent.com/9797472/133972781-60523b46-7958-4bcf-941d-fb2e9fb8ba81.png)

| Snippet Text  |  Description |
|---------------|---------------|
|`#if` |  if region directing
|`#region` |  region directive
|`checked` |  checked statement
|`class` |  define new class
|`do` |  do…while loop
|`else` |  else statement
|`enum` |  define a new enum
|`for`|  for loop
|`foreach`|  foreach block
|`forr`|  Reverse for loop
|`if`|  if statement
|`interface`|  define a new interface
|`lock`|  lock statement
|`namespace`| define a new namespace |
|`struct`| define a new struct
|`try`|  try catch
|`tryf`|  try finally
|`unchecked`|  unchecked block
|`unsafe`|  unsafe block
|`using`| using statement
|`while`|  while loop
</details>

### Visual Basic Snippet Behavior in Visual Studio

The VB snippet design is different they ship about 100 snippets that offer a more expansive set of capabilities _however_, they do not show them in the completion list and it is not expected that the user types out text such as `cw` or `prop` and gets the text expanded. Instead the user is expected to invoke the snippet menu (Ctrl+K+X).

### Visual Studio Snippet Capabilities

<details><summary>Some relevant details pulled from docs</summary>

*See the snipper schema docs [here](https://docs.microsoft.com/visualstudio/ide/code-snippets-schema-reference) for more information.*
There are two kinds of snippets that exist today in Visual Studio

- Surrounds With: surrounds a users selection
- Expansion: text expansion (i.e. cw -> Console.WriteLine();)

These VS snippets are allowed to have the following pre-defined variables

- `$selected$` is a predefined variable. It represents the text that was selected in the editor before invoking the snippet. The placement of this variable determines where the selected text appears in the code snippet that surrounds that selection.
- `$end$` is a predefined variable. When the user presses Enter to finish editing the code snippet fields, this variable determines where the caret (^) is moved to.

You can also define your own variables in the VS snippet engine and give them a default text but allow the user to tab to them and edit them on snippet insertion.

Each snippets can also have an [Assembly element](https://docs.microsoft.com/visualstudio/ide/code-snippets-schema-reference#assembly-element) indicating that an `<AssemblyReference>` should be added to the project file if not present and an [Import element](https://docs.microsoft.com/visualstudio/ide/code-snippets-schema-reference#import-element) that adds usings to the top of the file if not present.
</details>

### VS Code Snippet Capabilities

<details><summary>Some relevant details pulled from docs</summary>

*For more information see the docs [here](https://code.visualstudio.com/docs/editor/userdefinedsnippets#_create-your-own-snippets)*

VS Code's snippet system is similar to Visual Studio's in that it has a "trigger phrase" (such as `for`) that it uses as the decider for when to begin expansion, however its text transformation capabilities are a more dynamic as it allows regular expressions to be run over the output of a snippet.

The EBNF form of the regex syntax is defined as follows:
```bnf
any         ::= tabstop | placeholder | choice | variable | text
tabstop     ::= '$' int
                | '${' int '}'
                | '${' int  transform '}'
placeholder ::= '${' int ':' any '}'
choice      ::= '${' int '|' text (',' text)* '|}'
variable    ::= '$' var | '${' var '}'
                | '${' var ':' any '}'
                | '${' var transform '}'
transform   ::= '/' regex '/' (format | text)+ '/' options
format      ::= '$' int | '${' int '}'
                | '${' int ':' '/upcase' | '/downcase' | '/capitalize' | '/camelcase' | '/pascalcase' '}'
                | '${' int ':+' if '}'
                | '${' int ':?' if ':' else '}'
                | '${' int ':-' else '}' | '${' int ':' else '}'
regex       ::= JavaScript Regular Expression value (ctor-string)
options     ::= JavaScript Regular Expression option (ctor-options)
var         ::= [_a-zA-Z] [_a-zA-Z0-9]*
int         ::= [0-9]+
text        ::= .*
```

 They have the following predefined variables:

- `TM_SELECTED_TEXT` The currently selected text or the empty string
- `TM_CURRENT_LINE` The contents of the current line
- `TM_CURRENT_WORD` The contents of the word under cursor or the empty string
- `TM_LINE_INDEX` The zero-index based line number
- `TM_LINE_NUMBER` The one-index based line number
- `TM_FILENAME The` filename of the current document
- `TM_FILENAME_BASE` The filename of the current document without its extensions
- `TM_DIRECTORY` The directory of the current document
- `TM_FILEPATH` The full file path of the current document
- `RELATIVE_FILEPATH` The relative (to the opened workspace or folder) file path of the current document
- `CLIPBOARD` The contents of your clipboard
- `WORKSPACE_NAME` The name of the opened workspace or folder
- `WORKSPACE_FOLDER` The path of the opened workspace or folde

| Example |  Output | Explanation |
|---------|---------|-------------|
|`${TM_FILENAME/[\\.]/_/}` | 	example-123_456-TEST.js |	Replace the first . with _
|`${TM_FILENAME/[\\.-]/_/g}` |	example_123_456_TEST_js |	Replace each . or - with _
|`${TM_FILENAME/(.*)/${1:/upcase}/}`|	EXAMPLE-123.456-TEST.JS |	Change to all uppercase
|`${TM_FILENAME/[^0-9^a-z]//gi}`|	example123456TESTjs |	Remove non-alphanumeric characters
</details>

## User Experiences 

<details><summary>1.1	Snippet Discovery </summary>

A user starts typing `new` so they can start writing a new constructor. They are unaware of the built-in snippets Visual Studio provides but start typing the code for what they are trying to accomplish: 
 
![image](https://user-images.githubusercontent.com/46729679/165189198-a4dcb62a-57f8-46a4-837b-2a21b61bfb3a.png)

We will show the snippet for a constructer as soon as the type `new` as it is on our list of synonyms for that snippet:

![image](https://user-images.githubusercontent.com/46729679/165189214-ba69604a-f8cb-46a0-8d5b-d55a0af78beb.png)

If the user commits the snippet, it is persisted to the buffer as the “ghost text” preview had shown: 

![image](https://user-images.githubusercontent.com/46729679/165189238-d936d5a0-d9d7-44b1-b8a5-a3bd3e045fff.png)

If the user dismisses the completion list then the editor is placed back as it was with `new` being left alone:
 
![image](https://user-images.githubusercontent.com/46729679/165189254-b3e2d16b-98ca-41a5-8e33-7f5b0a61986a.png)

</details>

<details><summary>1.2	Using an Existing Snippet</summary>

A user types `ctor` (the text expansion for an existing snippet into the editor):

![image](https://user-images.githubusercontent.com/46729679/165189267-174a7c20-a056-4782-b04b-21bf5d9577bb.png)

When completion comes up, we will do custom filtering to: A. Select the `new constructor` completion item even though none of the text they’ve entered matches the text in this item. B. Use the `ghost text` API to show the user what will happen if they commit the completion item.

![image](https://user-images.githubusercontent.com/46729679/165189285-d05e751f-4187-4635-839f-0d1dac796c70.png)
 
If the user commits the snippet, it is persisted to the buffer as the `ghost text` preview had shown:

![image](https://user-images.githubusercontent.com/46729679/165189302-25fb92fa-ea0a-4639-b673-9da4d00f5c10.png)

If the user dismissed the completion list then the editor is placed back as it was with `ctor` being left alone:

![image](https://user-images.githubusercontent.com/46729679/165189320-c8410f37-81a2-496f-95bf-5daaa22122de.png)

</details>

<details><summary>1.3	Semantic Awareness</summary>

Today the editor shows and commits snippets even if they do not make sense. Here the developer is being suggested the “for” snippet even though it can never be valid inside of a type: 

![image](https://user-images.githubusercontent.com/46729679/165189343-9938fae4-d66b-4797-9a67-bdcd27ea56db.png)

In the new experience snippets will be aware of the context they are appearing in and only show if
they are reasonable. Notice we will now suggest `FormattableString` as a completion item assuming the
user is writing out a new property or field in this type:

![image](https://user-images.githubusercontent.com/46729679/165189386-bfd735f9-b0c9-466a-9e67-4c6b30ab6e5a.png)

However if the user is in a situation where this snippet makes sense we will: A. Filter to it in the completion list based on hard-coded synonyms. B. Show a `ghost text` preview of what the committed code will look like. C. Populate the snippet based on the semantic context in which it appears (notice how `points.Length` is automatically used in the snippet).

![image](https://user-images.githubusercontent.com/46729679/165189399-80c2be20-b06f-4128-96d0-c592d06b9a22.png)

If the user commits the snippet, we will use the existing snippet expansion experience that Visual Studio offers today to allow the user to fill in the `placeholders` that are automatically generated for them:

![image](https://user-images.githubusercontent.com/46729679/165189406-7dc74d3f-791b-43fc-b89a-16da08d4efb6.png)

If the user dismisses completion, they are left with the same contents in their buffer as they had before:

![image](https://user-images.githubusercontent.com/46729679/165189427-70ce5cd3-c498-47b8-9fd1-52534fea11c3.png)

</details>

----

## Tracking Issues

- [ ] [https://github.com/dotnet/roslyn/issues/5574](https://github.com/dotnet/roslyn/issues/5574)
- [ ] [https://github.com/dotnet/roslyn/issues/60628](https://github.com/dotnet/roslyn/issues/60628)
- [x] [https://github.com/dotnet/roslyn/pull/60860](https://github.com/dotnet/roslyn/pull/60860)
- [x] [https://github.com/dotnet/roslyn/pull/57346](https://github.com/dotnet/roslyn/pull/57346)
- [X] [https://github.com/dotnet/roslyn/pull/59303](https://github.com/dotnet/roslyn/pull/59303)

**Comment by JoeRobich** [comment]
- Questions:
  - Should we wait for new Editor snippet experience?
    - Is a new snippet experience on the roadmap?
    - Would it cover our expected scenarios?
  - Will we work with the VS for Mac snippet engine?
  - Will single tab insert these new snippets?
    - Argument completion is adding an option for it to trigger on single tab, snippets could inherit this option/behavior.
    - 3rd party snippets implemented by completion provider would need special consideration to trigger the snippet behavior.
  - Will this clash with IntelliCode and whole line completion?

Proposal: Update current snippet provider to show description text in completion list. Validate the idea by porting the most used snippets.

**Comment by KirillOsenkov** [comment]
One thing that I don't think we can break is the presence of snippets in the completion list. Typing `foreach + TAB + TAB` is very important muscle memory for many users. One thing we can do better is only show snippets that make sense in the context, e.g. foreach is a statement so don't show it in the expression context. If users don't see `foreach` in completion it might cause confusion.

Also another important aspect of the experience is TAB TAB (snippet insertion) can create several fields and linked fields (represented as tags in text) that you can tab through, edit and finally commit using Enter. VB snippets work differently.

Also note that some snippets have special sauce, e.g. `ctor`, and switch for enums will expand the enum fields.

Two invocation shortcuts are also important: Ctrl+K,X (Insert Snippet) and Ctrl+K,S (Surround With). Surround With is heavily used, Insert Snippet perhaps not as much.

From the implementation perspective, I think if you get your prototype working in VS Windows, it should be relatively straightforward to make it work in VSMac as well. I anticipate no significant hurdles there. So I'd say when a preview of the working functionality is ready, we can take another look and try to make it work in VSMac as well.

**Comment by jmarolf** [comment]
@mikadumont the merged spec can be found here: https://github.com/dotnet/roslyn/blob/feature/semantic-snippets/docs/ide/specs/semantic_snippets.md

**Comment by jmarolf** [comment]
related to https://github.com/dotnet/roslyn/issues/5574

**Comment by tamlin-mike** [comment]
For the sake of sanity, please have a look at a system where this already works, and works well. Just about any editor produced by Jetbrains (basically the gold standard for code editors) would probably be the quality to aim for.

The XML mess in VS is a horror show. It's enough to make you cry. The snippets description language is at the same time so complex and convoluted as to make it effectively impossible to author, understand, or maintain, and so primitive in its expressiveness it's close to useless for anything but the Microsoft-provided tiny snippets (though even they are so frequently screwing up formatting of the code and whatnot that at least I rather stay away from them than cleaning up the mess after it). It's basically the software equivalent of a fermenting septic tank, and the best one could do would be to burn it, to make sure it never could raise from the dead.

So yeah, I'm 100% with you that this needs TLC, but for the love of god stay the hell away from XML.

Also, I see no reason to even _think_ about waiting for VS to do anything in this area. Judging by history it won't happen. Period. They don't care. At all. That's a verifiable fact. On the off chance anything did come out of VS (in only the next 5-15 years) it would with five nines probability be even worse than the current XML-based junk.

So just "do it", but do it right; make the design process of the snippets language an _open process_. Allow input from developers, the ones that would actually _use_ this, where many of them have decades more experience than you might have, some might have _domain specific_ experience, and some might even have _passion_ for _quality_.

Please try to aim for something that can be a _usable_ and _sustainable_ language for this, i.e. it should be _fit for purpose_, that can be used and reused for decades to come.

Just my 0.02.

**Comment by jmarolf** [comment]
# Visual Basic and Semantic Snippets

In [the spec](https://github.com/dotnet/roslyn/blob/features/semantic-snippets/docs/ide/specs/semantic_snippets.md) for snippets I wrote the following

> We are not changing the VB experience for snippets (yet)
> VB has a very different design than C#.

Which is true, but after some consideration (and gently prodding from @CyrusNajmabadi and others) I think we need to have a more firm plan than just "we'll figure it out later."

## Background

VB already has "Semantic Snippets" in many ways and when talking to VB developers they constantly asked "Doesn't VB already do that?" and in many ways they do.

From an implementation standpoint the VB editing experience does not have any snippets for keywords like `If` or `For` but **that doesn't matter because a superior experience is provided via the pretty printer**.

https://user-images.githubusercontent.com/9797472/187209831-48eaeff4-cfed-4a73-b0e8-5eb6d97710ae.mp4

The keyword recommender is already "sematically aware" as it provides keywords only in location where they make sense (one the the things we want to bring to C#).

https://user-images.githubusercontent.com/9797472/187209924-20113f38-106f-417b-9f6d-b2ee6b1030f7.mp4

The pretty printer also fixes up these keywords and expands them in a nice way. But the Visual Studio snippet engine is never involved.

https://user-images.githubusercontent.com/9797472/187209958-8ea1ba99-988b-4e98-bde3-15ddd2dfa15b.mp4

## Proposal

Based on this context I propose the following experience for Visual Basic

### 1. Visual Basic will not get any keyword-based Semantic Snippets

They keyword recommender + pretty printer already handles this case. In the future we could unify implementations with C# snippets if there is a valid reason to do so, but from an experience perspective nothing should change here for the Visual Basic programmer.

### 2. Visual Basic will get the same "Coding Task" snippets as C#

The following "Coding Task" snippets are offered today in C#. We should ship with the VB equivalents for these in Visual Basic.

- prop (declare new auto prop)
- propfull (declare property with backing field)
- propg (declare new auto prop with private setter)
- cw (`Console.WriteLine`)
- mbox (`MessageBox.Show("Test")`)
- sim (`Function Main(args As String()) As Integer`)
- svm (`Sub Main(args As String())`)
- attribute (create new attribute type)
- equals (implement equality for type)
- exception (create new exception type)
- indexer (create new indexer property)
- invoke (invoke event, correctly handling null)
- iterator (declare yield returning method)
- iterindex (declare custom iterator type)

(Note: @ankitbko is implementing these in order of most used, some of these may be out-of-date with modern programming and we may not end up implementing them for C#, the intent that any "Coding Task" snippet that is not a key word should also have Visual Basic support)

## 3. We do not want to add VB specific snippets

I do not think there is need for us to consider specific Visual Basic snippets in order to ship the Semantic Snippets feature. That doesn't mean we can't take contributions that add things only for VB, but the intent it that we keep the experiences, as best we can, consistent with each other.

## 4. If we end up shipping Semantic Snippets with the SDK they should work for both Visual Basic and C#

Slightly further away (and likely after we ship the initial version of Semantic Snippets) we want to provide an API for folks on the runtime and ASP.NET teams to write their own snippets and ship them in the .NET SDK like we do with analyzers. Like analyzers the expectation would be that these work for both C# and Visual Basic unless they apply to a language feature that only one language supports.

**Comment by jmarolf** [comment]
@KathleenDollard and @KlausLoeffelmann would appreciate any additional thoughts you may have on this.

**Comment by CyrusNajmabadi** [comment]
> They keyword recommender + pretty printer already handles this case. In the future we could unify implementations with C# snippets if there is a valid reason to do so, but from an experience perspective nothing should change here for the Visual Basic programmer.

How does the keyword recommender help with the `for/foreach` cases?

**Comment by JoeRobich** [comment]
Design Meeting Notes (2022-08-29):
**Resolution:**
- We should support Semantic Snippet in VB and ship "Coding Task" snippets for VB

**Comment by jmarolf** [comment]
> How does the keyword recommender help with the for/foreach cases?

1. the keyword recommender is semantically aware so it will only offer completion for valid locations. This is technically true for the C# keyword recommender as well but the snippet for `for`/`foreach` is always available so the end user experience for the past 10 years has been that `for`/`foreach` is always in the completion list, where as VB has only ever shown them via the keyword recommender, so they only appear if they are valid to insert.
2. The VB pretty printer then inserts the text for `for`/`foreach`. The pretty printer has a wide berth to do as it wishes here and always has. If there is some new form we want the pretty printer to insert then we should add that to the pretty printer.

**Comment by CyrusNajmabadi** [comment]
> The VB pretty printer then inserts the text for for/foreach. The pretty printer has a wide berth to do as it wishes here and always has. If there is some new form we want the pretty printer to insert then we should add that to the pretty printer.

I guess i don't really understand the distinction between "the pretty printer" and "Semantic snippets".  I view "semantic snippets" as the space of "smartly inserted code constructs with rich functionality attached to them".

**Comment by jmarolf** [comment]
I don't disagree. For a user experience perspective, they should be the same, just different implementations. I do not think we should _remove_ the pretty printer for VB and use snippets instead for keywords. That changes a lot of how the VB typing experience works. If there is some additional smarts we want to add, lets just add it to the pretty printer in VB.

**Comment by CyrusNajmabadi** [comment]
>  If there is some additional smarts we want to add, lets just add it to the pretty printer in VB.

we can do that, but i'm worried about differing behaviors.  For example, if @akhera99 adds support for 'for/foreach' in C#, i would want us to share all that code with VB so we have a uniform experience. If it's in a different component i worry we won't.

**Comment by neuecc** [comment]
In the latest Visual Studio, `cw` has become `Console.Out.WriteLineAsync` at async context,
but I think this is unnecessary.
It is a bad feature unless it is reverted back to `Console.WriteLine` or made configurable.

**Comment by CyrusNajmabadi** [comment]
WriteLineAsync is preferred in async contexts so you don't block the calling thread doing IO.

**Comment by neuecc** [comment]
ConsoleStream only has sync write.
https://source.dot.net/#System.Console/System/IO/ConsoleStream.cs,c0e0ba628fa4c224
UnixConsoleStream:
https://source.dot.net/#System.Console/System/ConsolePal.Unix.ConsoleStream.cs,13bf065fbb4646d5
WindowsConsoleStream:
https://github.com/dotnet/runtime/blob/f9f7f8079377039a8d42269434a41ca55d14d591/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L1076

In other words, I don't think Async makes sense.

**Comment by CyrusNajmabadi** [comment]
@neuecc : https://learn.microsoft.com/en-us/dotnet/api/system.io.textwriter.writelineasync?view=net-7.0

**Comment by neuecc** [comment]
Console.Out creates
`new StreamWrite(ConsolePal.OpenStandardOutput())`
Eventually, `StreamWriter(TextWriter)` calls `WriteAsync` and `FlushAsync` of the internal Stream.
If they are not overridden, they are in a "sync over async" state on the ThreadPool, simulating asynchronous operation.

**Comment by CyrusNajmabadi** [comment]
Feel free to give the runtime feedback if you feel that WriteLineAsync will have problems on console streams.  In the worst case, you will have an impl that is as bad as WriteLine would be.  In hte best case, you have an impl that is a good async citizen.  The snippets encode best practices for async code to ensure that you're doing the right hting, and you can benefit from any improvements the runtime makes now and in the future.  Emitting WriteLine would make that not possible.

**Comment by neuecc** [comment]
Will there ever be a truly asynchronous implementation of ConsoleOut? If not, we should revise the snippet after such an implementation is established. As it stands, we're implementing a snippet that performs worse than WriteLine and is less user-friendly.

However, if we're concerned about performance, we should avoid writing to the console every time in the first place.

**Comment by KirillOsenkov** [comment]
I’m more worried about breaking people’s muscle memory. Why not add ‘cwa’ for the Async version and keep ‘cw’ the classic version.

Anecdotally, 100% of usages I’m seeing is console apps that are sync, as well as small throwaway Program.cs programs and tools. I’ve never seen Console.WriteLine in an async context.

**Comment by CyrusNajmabadi** [comment]
> Will there ever be a truly asynchronous implementation of ConsoleOut? 

I don't see any reason why not. Esp. as many OS' are happy to provide async io. 

> Why not add ‘cwa’ for the Async version and keep ‘cw’ the classic version.

The *very* intentional idea here was that you just use one snippet and you get the best emit per context. It's not on you to remember which to use and when, the system figures it out for you.

**Comment by CyrusNajmabadi** [comment]
> Anecdotally, 100% of usages I’m seeing is console apps that are sync, as well as small throwaway Program.cs programs and tools. I’ve never seen Console.WriteLine in an async context.

If you have a sync app, then we'll emit C.WL.  We only emit WLA if it's an async context. :-)

**Comment by CyrusNajmabadi** [comment]
> As it stands, we're implementing a snippet that performs worse than WriteLine and is less user-friendly.

Why would it perform worse than WL?

**Comment by CyrusNajmabadi** [comment]
> However, if we're concerned about performance, we should avoid writing to the console every time in the first place.

I see no relation between those ideas. You can want to write to the console and still not block a thread.  Perf and async IO go hand in hand.

**Comment by neuecc** [comment]
Sync over Async is a BAD PRACTICE for async, right?
I am saying that is the situation.

**Comment by neuecc** [comment]
Note that using ConsoleWrite for StandardOut is slow, so I have created my own library to handle this.
https://github.com/Cysharp/ZLogger

**Comment by mayuki** [comment]
I understand that it is generally preferred to use async methods within async contexts, and I think that the ability to switch snippets according to the context is great.

However, when it comes to `Console`, its implementation is not asynchronous on many platforms, and making it asynchronous incurs additional costs (`Task` allocation, etc.). Many use cases of `Console.WriteLine` simply display messages or are used for `printf` debugging and are not directly used in cases where performance is required.

Above all, `Console.WriteLine` is used as an idiom called `Cosnole.WriteLine` regardless of whether it is in an async context or not.

For example, if I write a simple console tool with top-level statements and have a line like `await new HttpClient().GetStringAsync("http://www.google.com");`, it will become an async context, but it seems unfamiliar to have `Console.Out.WriteLineAsync` after that. (It doesn’t happen now.)

I think it would be best to keep them separate as `cw` and `cwa`.

**Comment by MisinformedDNA** [comment]
If `Console.Out.WriteLineAsync` is always sync over async, then there is no benefit to it. @stephentoub has spoken out against the practice.

**Comment by stephentoub** [comment]
Console.Out is always a synchronized text writer (`TextWriter.Synchronized`), with all operations guarded by a Monitor; all operations that write to the console are guarded by the same lock, and any time we've tried to change that, we've had to revert because significant numbers of apps have taken dependencies on being able to do things like `lock (Console.Out) { Console.ForegroundColor = ...; ... multiple writes ...; Console.ResetColor(); }`.  Async operations on such a writer get converted into synchronous ones.  Using `await Console.Out.WriteLineAsync(...)` instead of `Console.WriteLine(...)` has no upside and only overhead.

**Comment by neuecc** [comment]
VS 17.8.0 Preview 2.0 is still `Console.Out.WriteLineAsync`.

**Comment by CyrusNajmabadi** [comment]
Unclear what is left here @akhera99 . Can yuo please update or close this out?

**Comment by akhera99** [comment]
Closing since the working list of snippets is being updated here: https://github.com/dotnet/roslyn/issues/64144
